feat: kinexon parser

manuba95 opened the pull request:
Hi, I added an initial version of a parser for position data in the Kinexon format. Work is still in progress. I'll let you know when it's ready for feedback.

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+    team_infos = {}
+    t = []
+    # check for group id
+    if parameter_links["group_id"] is not None:
```
the code block for both conditions is very similar. is there a way of inserting the artificial 0 group within `_get_parameter_links` and have no distinction here?

manuba95 replied:
I am not entirely happy with this part myself. However, I don't think I can solve it in `_get_parameter_links` as this distinction directly influences the structure of the `team_infos` that is created here. One way to "fix" this would be to make the distiction inside the loop. However, this would only get rid of the loop itself. Additionally it would add an unnecessary if statement at each iteration, thus, lowering performance. 
But I agree that this should be restructured somehow and will come back to it!

manuba95 replied:
Done

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+                    break
+                line = line_string.split(",")
+                # skip the header of the file
+                if line[parameter_links["time"]].isdecimal():
```
```suggestion
                if not line[parameter_links["time"]].isdecimal():
                   break
```

To avoid nesting?

manuba95 replied:
I'm not sure I understand your intention. However, this change would lead to a break in the first iteration as it contains the header of the file. and `line[parameter_links["time"]]` is `"ts in ms"`.  The entire point of the condition is to skip this one line. If the same can be done without nesting I'm all for it.

manuba95 replied:
Fixed it and reduced nesting by two layers

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+    timestamps.sort()
+    timestamps = np.array(timestamps)
+    minimum_time_step = np.min(np.diff(timestamps))
+    frame_rate = int(1000 / minimum_time_step)
```
If the framerate, for whatever reason, ever turned out to be non-integer, this would automatically round it without notification. 

```
if not frame_rate.is_integer():
   warnings.warn(f"Non-integer frame rate: {frame_rate} detected, round to ....

draabe replied:
Whats the 1000? I assume you go from milliseconds to seconds (or something along this way). This might be one of the few cases where a magic number can be justified in the middle of the code, but that should be specified in a comment.

manuba95 replied:
Done

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+
+    Parameters
+    ----------
+    team_infos: Dict[str, Dict[str, List[str]]],
```
as above, this needs clarification

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+    return positions
+
+
+#
```
This should be removed in the final version

manuba95 replied:
Done

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+    Returns
+    -------
+    positions: XY
+        XY-Object for the whole game. The order of groups is ascending according to
```
This is not clear to me, do you intend to pass a list of XY, or do you concatenate the data?

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+                    break
+                line = line_string.split(",")
+                # skip the header of the file
+                if line[parameter_links["time"]].isdecimal():
```
as above

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+                break
+            line = line_string.split(",")
+            # skip the header of the file
+            if line[parameter_links["time"]].isdecimal():
```
as above

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+                if y_coordinate != "":
+                    xydata[group][row, y_col] = y_coordinate
+
+    # join group arrays
```
I see now. We're changing this anyway soon, but for now maybe return a list of XY objects

manuba95 replied:
Returning list of XY objects for now

draabe reviewed (changes requested):
Hi @manuba95, thanks a lot for your contribution! I think that's a solid approach to parsing the Kinexon files with all the challenges attached. I've made some comments in the code, and I have a few general remarks:

- All the helper functions and dictionaries are obviously necessary to deal with the complexity of unknown column names. The variable name choices however where not entirely clear to me, and maybe a little cosmetics here could help making those parts even more accessible. For example, I struggled with "parameters" for column names, maybe just "columns" will do?
- Although it's "frame rate" in English, we use "framerate" as a variable name instead of "frame_rate" throughout the code and in the docs, so I'd suggest to change that
- We also use "teams" as a variable to denote all teams in an observation throughout the code and in the docs, maybe its possible to change "groups" to "teams" where applicable?
- Other variable names I found confusing were team_info and the name of links. I've just added a section in the docs on IDs and links, maybe you can derive more concise names from there?
- I think the docstrings could use a little more info on how the methods work, as there is more user choices to be made in this parser compared to the others. Maybe add a little Notes section and explain what it means to choose between mapped_id or sensor_id and what the defaults/resolution order is.

Otherwise, I think it's a great addition, and with these minor points worked on it should definitely be merged!

manuba95 commented on docs/source/modules/io/kinexon.rst:
```diff
@@ -0,0 +1,6 @@
+====================
```
done

manuba95 commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+
+
+def get_column_names_from_csv(filepath_data: Union[str, Path]) -> List[str]:
+    """Reads first line of a Kinexon and extracts the column names.
```
changed it to "Kinexon.csv-file" to be consistent with other docs

manuba95 commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+    filepath_data: Union[str, Path]
+) -> Tuple[Dict[str, Dict[str, List[str]]], int, int, int]:
+    """Reads Kinexon's position data file and extracts meta-data about groups, sensors,
+    length and framerate
```
Done

manuba95 commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+
+    Parameters
+    ----------
+    filepath_data: str of pathlib.Path
```
Done

manuba95 commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,349 @@
+
+    parameter_links = _get_parameter_links(str(filepath_data))
+
+    # sensor-identifier
```
I don't quite understand what you mean by "here". If you are referencing the `sensor_links`, I remove all columns that don't contain ´pID´s. What part of it needs more detailed documentation?

manuba95 commented:
Hi @draabe, thanks for the valuable comments to this contribution! I already addressed your comments in the code. Regarding your general remarks:

- changed parameter to column
- changed frame_rate to framerate
- I changed group to team for consistency. However, I want to leave it out for discussion if "team" is the appropriate term here. The "group id" in Kinexon files distincts any kind different "groups" of sensors, including the balls. I think we trade interpretability for consistency if Balls or any kind of distinct sensor groups are considered to be in a "team". 
- Added terminology from the docs
- Explained optional arguments in the docs

draabe commented:
You are absolutely right, I forgot that the Ball is included. I think we should follow your suggestion and stay with groups.

I'll review the rest of your changes asap

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+    filepath_data: str of pathlib.Path
+        Full path to Kinexon.csv-file.
+
+
```
redundant blank line

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+            - group_id: 'group id'
+            - x_coord: 'x in m'
+            - y_coord: 'y in m'
+
```
redundant blank line

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+    pID_dict: Dict[str, Dict[str, List[str]]],
+        Nested dictionary that stores information about the pIDs from every player-
+        identifying column in every group.
+        'pID_dict[group_identifier][identifying_column] = [pID1, pID, ..., pIDn]'
```
```suggestion
        'pID_dict[group_identifier][identifying_column] = [pID1, pID2, ..., pIDn]'
```

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+    pID_dict: Dict[str, Dict[str, List[str]]],
+        Nested dictionary that stores information about the pIDs from every player-
+        identifying column in every group.
+        'pID_dict[group][identifying_column] = [pID1, pID, ..., pIDn]'
```
```suggestion
        'pID_dict[group][identifying_column] = [pID1, pID2, ..., pIDn]'
```

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+                break
+
+    links = {}
+    for gID in pID_dict:
```
what is gID? group_ID?

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+
+    Parameters
+    ----------
+    filepath_data: str of pathlib.Path
```
```suggestion
    filepath_data: str or pathlib.Path
```

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+
+    Parameters
+    ----------
+    filepath_data: str of pathlib.Path
```
```suggestion
    filepath_data: str or pathlib.Path
```

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+    Returns
+    -------
+    positions: List[XY]
+        XY-Object for the whole game. The order of groups is ascending according to
```
```suggestion
        List of XY-objects for the whole game, one per group. The order of groups is ascending according to
```

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+    positions: List[XY]
+        XY-Object for the whole game. The order of groups is ascending according to
+        their group_id. The order inside the groups is ascending according to their
+        appearance in the data.
```
```suggestion
        appearance in the data. If no groups are specified in the file, all data is gets assigned to a dummy group "0".
```

draabe commented on floodlight/io/kinexon.py:
```diff
@@ -0,0 +1,359 @@
+    # no groups
+    else:
+
+        def get_group_id(single_line):
```
Although I like your solution to dynmaic group id creation, nested functions are not the most stable way. I've also noticed you defined `get_group_id` four times throughout your code. Maybe there is a way to refactor all four into one top level function `_get_group_id(single_line)` that returns `single_line[column_links[group_identifier]]` if the value exists, or "0" otherwise.

If I'm missing something and thats not straightforward to solve, I'm fine going with your solution!

draabe reviewed (changes requested):
Hi, code looks real great now! Just noted a few typos and made one suggestion, but thats optional.

manuba95 commented:
I fixed the typos, removed the nested functions and added some logic to the use of available sensor identifiers. I'm fine with a merge now.

draabe reviewed (approved):
Looks great!
